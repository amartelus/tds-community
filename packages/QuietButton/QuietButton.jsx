import React from 'react'
import styled from 'styled-components'
import { small } from '@tds/shared-typography'
import { borders } from '@tds/shared-styles'
import PropTypes from 'prop-types'

const baseButton = {
  boxSizing: 'border-box',
  margin: '3px',
  padding: '0 2rem',
  cursor: 'pointer',
  background: '#FFFFFF',
  transition: 'all 0.2s ease-in-out',
  display: 'flex',
  flexDirection: 'row',
  justifyContent: 'space-between',
  minWidth: '44px',
  border: '1px solid #71757B',
  position: 'relative',
  borderRadius: '3px',
  color: '#2A2C2E',

  '@media (prefers-reduced-motion: reduce)': {
    transition: 'none !important',
  },
}

const StyledQuietButton = styled.button(baseButton, small, borders.rounded, {
  '&:hover': {
    border: '3px solid #71757B',
    midWidth: '72px',
  },
  '&:active': {
    border: '3px solid #54595F',
    background: '#D8D8D8',
  },
  '&:focus': {
    background: '#FFFFFF',
    border: '2px solid red',
    outline: '3px solid blue',
    '&:before': {
      position: 'absolute',
      content: '',
      zIndex: -1,
      border: '3px solid green',
    },
  },
})

/**
 * The quiet button is used for optional actions, and only comes in one variant and size
 * @version ./package.json
 */

const QuietButton = props => {
  // const noDisabled = preventDisabling()
  return <StyledQuietButton>{props.children}</StyledQuietButton>
}
// check with Jesse - 'we will not be programmatically enforcing child types'
// so do we need this propTypes object?
// how to include interactive icons on this list?
//
QuietButton.propTypes = {
  /**
   * Button text.
   */
  // this is allowing children of any type essentially
  children: PropTypes.node.isRequired,
}
export default QuietButton

// Focus: solid white fill #FFFFFF

// first border: 2px white #FFFFFF (inside)

// second border: 3px  Raven grey #71757B (outside)

// third border: 3px Shuttle grey #54595F(inside)

// children: or([
//   PropTypes.string,
//   componentWithName('A11yContent'),
//   componentWithName('InteractiveIcon'),
// ]).isRequired
