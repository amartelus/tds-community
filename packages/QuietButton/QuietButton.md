Quiet buttons are used sparingly and for an optional action. Mainly used in internal applications like My TELUS (web and app) or Casa; it only comes in one variant and size.

### Usage criteria

- Use for interactions that are not necessarily mandatory
- Use for enhanced experience where if the button does not exist, the customer can still do what they need to
- Keep the text short and precise, recommendation is text that describes an action (Ie. Copy, Edit, Print)
- Make use of the A11yContent core component to provide more written context for assistive technology users for when they navigate a page using only button landmarks
- Buttons should not be disabled

#### QuietButton with Icons

- Only dependent icons can be added into buttons, they can be on the left or right of the label
- The colour of the text and icon have to be the same

```jsx
<QuietButton>Hello World!</QuietButton>
```
