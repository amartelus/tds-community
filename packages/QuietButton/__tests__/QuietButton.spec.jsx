import React from 'react'
import { shallow, mount } from 'enzyme'
import 'jest-styled-components'

import QuietButton from '../QuietButton'

describe('QuietButton', () => {
  const doShallow = (props = {}) => shallow(<QuietButton {...props} />)
  const doMount = (props = {}) => mount(<QuietButton {...props} />)

  it('renders', () => {
    const quietButton = doShallow()

    expect(quietButton).toMatchSnapshot()
  })

  // from Modal testing file, why mount vs shallow not completely sure
  it('renders', () => {
    const quietButton = doMount()

    expect(quietButton).toMatchSnapshot()
  })

  it('does other things', () => {
    const quietButton = doMount()

    expect(quietButton).toExist()
  })
})

// mounting: Full DOM rendering including child components
// shallow: Renders only the single component, not including its children. This is useful to isolate the component for pure unit testing.
// render: no React life cycle, Less costly than mount but provides less functionality

// Always begin with shallow
// If componentDidMount or componentDidUpdate should be tested, use mount
// If you want to test component lifecycle and children behavior, use mount
// If you want to test children rendering with less overhead than mount and you are not interested in lifecycle methods, use render
