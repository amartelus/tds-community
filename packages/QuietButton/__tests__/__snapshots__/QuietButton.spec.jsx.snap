// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuietButton renders 1`] = `
<QuietButton__StyledQuietButton>
  <Dependent />
  Words
</QuietButton__StyledQuietButton>
`;
